src/sentry/release_health/duplex.py

```diff
@@ -761,43 +761,67 @@ def _org_from_id(org_id: OrganizationId) -> Organization:
         rollup: Optional[int],
         organization: Optional[Organization],
         schema: Optional[Schema],
         *args: Any,
     ) -> ReleaseHealthResult:
+        # Note: having both featureflags enabled at once is deadly for endpoint
+        # performance, but does work.
+        should_return_metrics = organization is not None and features.has(
+            "organizations:release-health-return-metrics", organization
+        )
+        should_check_metrics = organization is not None and features.has(
+            "organizations:release-health-check-metrics", organization
+        )
+        if not isinstance(should_compare, bool):
+            try:
+                should_compare = should_compare()
+            except Exception:
+                capture_exception()
+                should_compare = False
+
+            if not should_compare:
+                # if should_compare=False, then we already know the metrics result is bad and shouldn't return it.
+                should_return_metrics = False
+
         sessions_fn = getattr(self.sessions, fn_name)
+        metrics_fn = getattr(self.metrics, fn_name)
+
+        sessions_result = None  # get rid of unbound warnings -- this line shouldn't be necessary
 
         now = datetime.now(pytz.utc)
         tags = {"method": fn_name, "rollup": str(rollup)}
-        with timer("releasehealth.sessions.duration", tags=tags, sample_rate=1.0):
-            ret_val = sessions_fn(*args)
-
-        if organization is not None and features.has(
-            "organizations:release-health-check-metrics", organization
-        ):
-            if not isinstance(should_compare, bool):
-                # should compare depends on the session result
-                # evaluate it now
-                should_compare = should_compare(ret_val)
+        incr(
+            "releasehealth.metrics.should_return",
+            tags={"should_return": str(should_return_metrics), **tags},
+        )
+        if should_check_metrics or not should_return_metrics:
+            with timer("releasehealth.sessions.duration", tags=tags, sample_rate=1.0):
+                sessions_result = sessions_fn(*args)
 
+        if should_check_metrics:
             try:
                 run_comparison.delay(
                     fn_name,
-                    getattr(self.metrics, fn_name),
+                    metrics_fn,
                     should_compare,
                     rollup,
                     organization,
                     schema,
                     function_args=args,
-                    sessions_result=ret_val,
+                    sessions_result=sessions_result,
                     sessions_time=now,
                 )
             except Exception:
                 capture_exception()
 
-        return ret_val
+        if should_return_metrics:
+            with timer("releasehealth.metrics.duration", tags=tags, sample_rate=1.0):
+                return metrics_fn(*args)
```

Comment:
What if we move this block to before the `run_comparison.delay`, and change `run_comparison` to accept either `metrics_fn` or `metrics_result`? Just so we don't end up running metrics queries twice.